labrador/src/jl_projection.rs

```diff
@@ -0,0 +1,91 @@
+use lark_algebra::{
+    ConfigRingGoldilocks256, ConfigZZpX, ConfigZZpXGoldilocks256, Goldilocks, Polynomial,
+    RingGoldilock256,
+};
+use rand::distributions::{Distribution, WeightedIndex};
+use rand::Rng;
+use std::ops::Neg;
+
+struct PiDistribution;
+
+impl Distribution<i8> for PiDistribution {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> i8 {
+        let values = [-1, 0, 1];
+        let weights = [1, 2, 1];
+        let dist = WeightedIndex::new(&weights).unwrap();
+        values[dist.sample(rng)]
+    }
+}
+
+pub fn jl_projection(
+    data: Vec<Vec<RingGoldilock256>>,
+    target_dim: usize,
+) -> Vec<Vec<RingGoldilock256>> {
+    let original_dim = ConfigRingGoldilocks256::DIM;
+    let mut projection_matrix: Vec<Vec<RingGoldilock256>> = Vec::with_capacity(target_dim);
+    for _ in 0..target_dim {
+        let mut row: Vec<RingGoldilock256> = Vec::with_capacity(original_dim);
+        for _ in 0..original_dim {
+            let mut poly: Vec<Goldilocks> = Vec::with_capacity(ConfigZZpXGoldilocks256::DIM);
+            for _ in 0..ConfigZZpXGoldilocks256::DIM {
+                let cell = match PiDistribution.sample(&mut rand::thread_rng()) {
+                    -1 => Goldilocks::from(1u64).neg(),
+                    0 => Goldilocks::from(0u64),
+                    1 => Goldilocks::from(1u64),
+                    _ => unreachable!(),
+                };
+                poly.push(cell);
+            }
+            row.push(RingGoldilock256::from_coefficients_vec_unchecked(poly));
+        }
+        projection_matrix.push(row);
+    }
+
+    let mut projected_data = Vec::with_capacity(data.len());
+    for data_vec in data.clone() {
+        let mut projected_vector: Vec<RingGoldilock256> = Vec::with_capacity(target_dim);
+        for projection_row in projection_matrix.clone() {
+            let projection_value = RingGoldilock256::dot_product(&data_vec, &projection_row);
+            projected_vector.push(projection_value);
+        }
+        projected_data.push(projected_vector);
+    }
```

Comment:
Probably using reference instead of cloning is better here. I am not very sure but we can discuss here

```suggestion
    let mut projected_data = Vec::with_capacity(data.len());
    for data_vec in &data {
        let mut projected_vector: Vec<RingGoldilock256> = Vec::with_capacity(target_dim);
        for projection_row in &projection_matrix {
            let projection_value = RingGoldilock256::dot_product(&data_vec, &projection_row);
            projected_vector.push(projection_value);
        }
        projected_data.push(projected_vector);
    }
```

Reply:
And I reckon the result of project should be a vector of pure numbers

for example, 

we have $r$ = 2 (the number of $\vec{s}$ ($\vec{s}=[s_1, s_2, ...]$ ))
$\vec{S_1} = [ [1,2], [3,4] ]$
$\vec{S_2} = [ [3,4], [5,6] ]$

lets the randomly sampled projection vector be

$\vec{\pi_1}^{(1)} = [1, 0, 1, -1]$
$\vec{\pi_2}^{(1)} = [0, 1, 1, 0]$

$p_1$ is calculated as below: 

$$
\begin{aligned}
p_1  &= <\vec{\pi_1}^{(1)}, \vec{s_1}> + <\vec{\pi_2}^{(1)}, \vec{s_2}> &= 1 \times 1 + 0\times2 + 1\times3 + (-1)\times4 + 0\times3 + 1\times4 + 1\times5 + 0\times6 &= 9
\end{aligned}
$$

I think in the end we will get a 256-length vector $\vec{p}$, which each element is a pure number, i.e., `[9, 2, 13, ...]`

Reply:
I think @FoodChain1028 is correct, refer to page 35 of  https://eprint.iacr.org/2024/311.pdf#page=34.19 , you will find the result of JL projection is in $Z_q^{2\lambda}$, which $\lambda = 128$
![image](https://github.com/user-attachments/assets/89154cc0-dac8-4b16-952e-a4ec7337266a)